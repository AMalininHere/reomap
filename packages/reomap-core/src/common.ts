import { tile2lat, tile2lng, lng2tile, lat2tile } from './utils/geo-fns';

export class LatLng {
  constructor(
    public readonly lat: number,
    public readonly lng: number,
  ) { }

  equals(other: LatLng) {
    return (
      this.lat === other.lat &&
      this.lng === other.lng
    );
  }
}

export class Point {
  constructor(
    public readonly x: number,
    public readonly y: number,
  ) { }
}

export const TILE_SIZE = 256;

export const absMinLatLng = new LatLng(
  tile2lat(Math.pow(2, 10), 10),
  tile2lng(0, 10)
);

export const absMaxLatLng = new LatLng(
  tile2lat(0, 10),
  tile2lng(Math.pow(2, 10), 10)
);

export function pixelToLatLng(width: number, height: number, zoom: number, center: LatLng, source: Point) {
  const pointDiffX = (source.x - width / 2) / TILE_SIZE;
  const pointDiffY = (source.y - height / 2) / TILE_SIZE;

  const tileX = lng2tile(center.lng, zoom) + pointDiffX;
  const tileY = lat2tile(center.lat, zoom) + pointDiffY;

  return new LatLng(
    Math.max(absMinLatLng.lat, Math.min(absMaxLatLng.lat, tile2lat(tileY, zoom))),
    Math.max(absMinLatLng.lng, Math.min(absMaxLatLng.lng, tile2lng(tileX, zoom)))
  );
}

export function latLngToPixel(width: number, height: number, zoom: number, center: LatLng, source: LatLng) {
  const tileCenterX = lng2tile(center.lng, zoom);
  const tileCenterY = lat2tile(center.lat, zoom);

  const tileX = lng2tile(source.lng, zoom);
  const tileY = lat2tile(source.lat, zoom);

  return new Point(
    (tileX - tileCenterX) * TILE_SIZE + width / 2,
    (tileY - tileCenterY) * TILE_SIZE + height / 2
  );
}
