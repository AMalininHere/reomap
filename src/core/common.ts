import { tile2lat, tile2lng, lng2tile, lat2tile } from './utils/geo-fns';
import { LatLng, Point } from './models';

export const absMinLatLng = new LatLng(
  tile2lat(Math.pow(2, 10), 10),
  tile2lng(0, 10)
);

export const absMaxLatLng = new LatLng(
  tile2lat(0, 10),
  tile2lng(Math.pow(2, 10), 10)
);

export function pixelToLatLng(width: number, height: number, zoom: number, center: LatLng, source: Point) {
  const pointDiffX = (source.x - width / 2) / 256.0;
  const pointDiffY = (source.y - height / 2) / 256.0;

  const tileX = lng2tile(center.lng, zoom) + pointDiffX;
  const tileY = lat2tile(center.lat, zoom) + pointDiffY;

  return new LatLng(
    Math.max(absMinLatLng.lat, Math.min(absMaxLatLng.lat, tile2lat(tileY, zoom))),
    Math.max(absMinLatLng.lng, Math.min(absMaxLatLng.lng, tile2lng(tileX, zoom)))
  );
}

export function latLngToPixel(width: number, height: number, zoom: number, center: LatLng, source: LatLng) {
  const tileCenterX = lng2tile(center.lng, zoom);
  const tileCenterY = lat2tile(center.lat, zoom);

  const tileX = lng2tile(source.lng, zoom);
  const tileY = lat2tile(source.lat, zoom);

  return new Point(
    (tileX - tileCenterX) * 256.0 + width / 2,
    (tileY - tileCenterY) * 256.0 + height / 2
  );
};
