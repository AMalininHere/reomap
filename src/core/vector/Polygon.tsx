import React from 'react';
import { useMapContext } from '../Context';
import { LatLng, Point } from '../models';

function makeSvgPath(points: Point[]) {
  return points
    .map((p, idx) => `${idx === 0 ? 'M' : 'L'} ${p.x} ${p.y}`)
    .join(' ') + 'z';
}

interface Props {
  positions: LatLng[][];
}

function Polygon(props: Props) {
  const {
    positions,
  } = props;

  const mapContext = useMapContext();

  const pathString = positions
    .map(subPositions => subPositions.map(mapContext.latLngToPixel))
    .map(makeSvgPath)
    .join(' ');


  return (
    <path fillRule="evenodd" fillOpacity="0.5" stroke="#555555" fill="#555555" strokeWidth={2} d={pathString} />
  );
}

export default React.memo(Polygon);
