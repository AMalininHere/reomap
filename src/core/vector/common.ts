import * as G from 'geojson';
import { LatLng, Point } from '../models';

export interface ElementProps<T extends G.Geometry> {
  geoElement: T;
  latLngToPixel: (ll: LatLng) => Point;
}