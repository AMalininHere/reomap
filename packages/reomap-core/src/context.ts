import React, { useContext } from 'react';
import { LatLng, Point, latLngToPixel, pixelToLatLng } from './common';

export class ContextState {
  constructor(
    public readonly center: LatLng,
    public readonly zoom: number,
    public readonly width: number,
    public readonly height: number,
  ) { }

  public readonly latLngToPixel = (source: LatLng) =>
    latLngToPixel(this.width, this.height, this.zoom, this.center, source);

  public readonly pixelToLatLng = (source: Point) =>
    pixelToLatLng(this.width, this.height, this.zoom, this.center, source);
}

const ctx = React.createContext<ContextState>(null!);

export const MapProvider = ctx.Provider;

export function useMapContext() {
  const data = useContext(ctx);

  return data;
}
