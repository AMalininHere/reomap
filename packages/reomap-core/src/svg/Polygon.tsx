import React, { Ref } from 'react';
import { useMapContext } from '../context';
import { LatLng, Point } from '../models';

function makeSvgPath(points: Point[]) {
  return points
    .map((p, idx) => `${idx === 0 ? 'M' : 'L'} ${p.x} ${p.y}`)
    .join(' ') + 'z';
}

interface Props {
  positions: LatLng[][];
}

type PathProps = Omit<React.SVGAttributes<SVGPathElement>, 'fillRule' | 'd'>;

function Polygon(props: Props & PathProps, ref: Ref<SVGPathElement>) {
  const {
    positions,
    ...pathProps
  } = props;

  const mapContext = useMapContext();

  const pathString = positions
    .map(subPositions => subPositions.map(mapContext.latLngToPixel))
    .map(makeSvgPath)
    .join(' ');

  const pathPropsWitdhDefaults: PathProps = {
    strokeWidth: 2,
    stroke: '#555555',
    fill: '#555555',
    fillOpacity: 0.5,
    ...pathProps
  };

  return (
    <path ref={ref} fillRule="evenodd" d={pathString} {...pathPropsWitdhDefaults} />
  );
}

export default React.memo(React.forwardRef(Polygon));
