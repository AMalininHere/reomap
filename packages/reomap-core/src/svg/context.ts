import { createContext, useContext } from 'react';
import { ContextState } from '../context';

const ctx = createContext<ContextState>(null!);
ctx.displayName = 'SvgLayerContext';

export const SvgLayerProvider = ctx.Provider;

export function useSvgLayerContext() {
  const data = useContext(ctx);

  return data;
}
