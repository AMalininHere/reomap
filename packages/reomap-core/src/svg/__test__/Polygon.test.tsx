import React from 'react';
import { render } from '@testing-library/react';
import Polygon from '../Polygon';
import { SvgLayerProvider } from '../context';
import { createContextState } from '../../context';
import { latLng } from '../../common';

describe('Polygon', () => {
  it('should render', () => {
    const contextState = createContextState(latLng(55.417, 85.276), 6, 800, 600);

    const geojsonPolygonPath: [number, number][][] = [
      [
        [
          80.57373046875,
          54.28446875235516
        ],
        [
          82.913818359375,
          52.522905940278065
        ],
        [
          87.835693359375,
          52.94201777829491
        ],
        [
          88.78051757812499,
          55.53484823078213
        ],
        [
          82.716064453125,
          56.806893398067466
        ],
        [
          80.57373046875,
          54.28446875235516
        ]
      ],
      [
        [
          82.41943359374999,
          54.17529672404642
        ],
        [
          82.353515625,
          55.015425940562984
        ],
        [
          83.507080078125,
          55.83214387781303
        ],
        [
          85.440673828125,
          55.88763544617004
        ],
        [
          87.725830078125,
          55.45394132943305
        ],
        [
          87.396240234375,
          53.42262754609993
        ],
        [
          85.05615234375,
          53.11381149316827
        ],
        [
          82.41943359374999,
          54.17529672404642
        ]
      ]
    ];

    const { container } = render(
      <svg>
        <SvgLayerProvider value={contextState}>
          <Polygon positions={geojsonPolygonPath.map(line => line.map(([lng, lat]) => latLng(lat, lng)))}/>
        </SvgLayerProvider>
      </svg>
    );

    expect(container).toMatchSnapshot();
  });
});
