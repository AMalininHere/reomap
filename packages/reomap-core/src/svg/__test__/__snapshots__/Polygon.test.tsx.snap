// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Polygon should render 1`] = `
<div>
  <svg>
    <path
      d="M186 390 L292 524 L516 493 L559 291 L283 187 L186 390 Z M270 398 L267 332 L319 267 L407 262 L511 297 L496 456 L390 480 L270 398 Z"
      fill="#555555"
      fill-opacity="0.5"
      fill-rule="evenodd"
      stroke="#555555"
      stroke-width="2"
    />
  </svg>
</div>
`;
